File: public/dsa/two_pointers/trapping_rain_water/prefix_arrays/trapping_rain_water.cpp
#include <bits/stdc++.h>
using namespace std;

int volumeOfTrappedRainWater(vector<int> &heights) {
    int n = heights.size();

    vector<int> maxOnLefts(n);
    vector<int> maxOnRights(n);

    maxOnLefts[0] = heights[0];
    for (int i = 1; i < n; i++) {
        maxOnLefts[i] = max(maxOnLefts[i - 1], heights[i]);
    }

    maxOnRights[n - 1] = heights[n - 1];
    for (int i = n - 2; i >= 0; i--) {
        maxOnRights[i] = max(maxOnRights[i + 1], heights[i]);
    }

    int totalRainWater = 0;
    for (int i = 0; i < n; i++) {
        int minOfLeftAndRight = min(maxOnLefts[i], maxOnRights[i]);
        int waterForThisBlock = minOfLeftAndRight - heights[i];
        totalRainWater += waterForThisBlock;
    }

    return totalRainWater;
}

// tests-start
int main() {
    vector<int> heights = {1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    int totalRainWater = volumeOfTrappedRainWater(heights);
    cout << totalRainWater << endl;
    return 0;
}
// tests-end

/*output
6
*/

File: public/dsa/two_pointers/trapping_rain_water/prefix_arrays/trapping_rain_water.py
from typing import List


class Solution:
    def volumeOfTrappedRainWater(self, heights: List[int]) -> int:
        n = len(heights)

        maxOnLefts = [0] * n
        maxOnRights = [0] * n

        maxOnLefts[0] = heights[0]
        for i in range(1, n):
            maxOnLefts[i] = max(maxOnLefts[i-1], heights[i])

        maxOnRights[n-1] = heights[n-1]
        for i in range(n-2, -1, -1):
            maxOnRights[i] = max(maxOnRights[i+1], heights[i])

        totalRainWater = 0

        for i in range(n):
            minOfLeftAndRight = min(maxOnLefts[i], maxOnRights[i])
            waterForThisBlock = minOfLeftAndRight - heights[i]
            totalRainWater += waterForThisBlock

        return totalRainWater


# tests-start
heights = [1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
sol = Solution()
print(sol.volumeOfTrappedRainWater(heights))
# tests-end

'''output
6
'''
